### DISCUSSION TITLE
Question: Step accuracy and microstepping

### DISCUSSION THREAD

**Discussion by OxygenLiu** [dzid26/StepperServoCAN#22] [General]
For a low cost 1.8 deg stepper motor, the step accuracy can be around +/- 1.5% (+/- 0.027deg) of full step. With the microstep of 256, the accuracy per microstep will be +/- 0.027 / (1.8/256) = +/- 384%. It may lead to some strange behavior under heavy load or near stall. 

I think up to 32 microsteps can be a reasonable choice for low cost steppers.

**Comment by dzid26** [comment]
With a close-loop commutation (stepper + calibrated encoder), the microstepping is not really a thing.  The angle used for activating the phases is interpolated and the more accuracy there the better. The commutation accuracy is up to 163microsteps (32768 encoder resolution / 200 steps).  

In practice, even if the encoder had less resolution it still work good. I think I tested with 2^11 resolution limit and I couldn't tell the difference in operation. That equates to 0.17deg error. But it makes sense right? Because that is only 10% of a full step - much bigger than 1.5%. Whether we command 90deg or 80deg electrical angle  (load angle), the generated torque will be similar. It only when we start increasing the error of load angle even further, the torque would change a lot.  
I don't think 1.5% or even 10% would account for #21.

**Comment by OxygenLiu** [comment]
Right, the step accuracy is irrelevant for the close loop control scheme.  

The 2^15 resolution results in an accuracy about `360 electrical deg / (32768/200steps) = 2.2 electrical deg`. As the electric load angle is set to 90 degree in the modified Park transform for Iq current, it may incorrectly trigger the `if sin < 0` and `if cos <0` statements to switch bridge state between 0 and 1. Under high current conditions, the coil inductance may lead to obvious current distortion and then torque ripple.

**Reply by dzid26** [reply]
Small correction.  360deg electrical angle consists of **four** "full steps". So theoretical resolution is 0.55deg. 
However in practice the sensor has a high frequency 3bit or more noise. It doesn't matter though as it gets lost when requesting PWM which has less resolution than the sensor. PWM timers have always tradeoff between better resolution and frequency.  (Roughly max PWM resolution is 72mhz/20khz=3600. 20khz due to audible noise.  It's a little more complicated for [vref](https://github.com/dzid26/StepperServoCAN/blob/main/firmware/src/BSP/A4950.h#L29) because there is hardware low pass filter there.)

Anyway, the switching between bridge 0 and 1 happens when current is already very small - close to zero (from Park transform). 
(Well, this is not always true during high speed, due to hardware delay in the aforementioned vref low pass filter.  But for low speed it is true. And in voltage-control it is always true because there is no hardware delay there.)


I really think the low speed torque ripple has to do with either magnetic saturation or encoder angle being [influenced](https://github.com/dzid26/StepperServoCAN/issues/21#issuecomment-1805665651) by high phase currents.
